File: work-ms2d-daemon/ms2d/clients/ms2_client_example.c
#include "ms2_client.h"
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char **argv)
{
    const char *socket_path = "/tmp/ms2d.sock";

    /* Allow override via command line */
    if (argc > 1) {
        socket_path = argv[1];
    }

    printf("MS2D C Client Example\n");
    printf("Connecting to daemon at: %s\n\n", socket_path);

    /* Connect to daemon */
    ms2_client_t *client = ms2_connect(socket_path);
    if (!client) {
        fprintf(stderr, "ERROR: Failed to connect to daemon\n");
        return 1;
    }
    printf("Connected successfully.\n\n");

    /* Get daemon status */
    int connected = 0;
    char signature[256] = {0};
    if (ms2_get_status(client, &connected, signature, sizeof(signature)) == 0) {
        printf("Daemon Status:\n");
        printf("  Connected to ECU: %s\n", connected ? "yes" : "no");
        printf("  ECU Signature: %s\n\n", signature);
    } else {
        fprintf(stderr, "ERROR: Failed to get daemon status\n");
    }

    /* Get single field value (RPM) */
    printf("Querying single field: rpm\n");
    double rpm = ms2_get_value(client, "rpm");
    printf("  rpm: %.1f\n\n", rpm);

    /* Get multiple field values */
    printf("Querying multiple fields: rpm, batteryVoltage, coolant\n");
    const char *fields[] = {"rpm", "batteryVoltage", "coolant"};
    double values[3] = {0};
    if (ms2_get_values(client, fields, 3, values) == 0) {
        printf("  rpm: %.1f\n", values[0]);
        printf("  batteryVoltage: %.2f V\n", values[1]);
        printf("  coolant: %.1f\n\n", values[2]);
    } else {
        fprintf(stderr, "ERROR: Failed to get multiple values\n");
    }

    /* List all available fields */
    printf("Listing all available fields:\n");
    int field_count = 0;
    char **field_names = ms2_list_fields(client, &field_count);
    if (field_names) {
        printf("  Total fields: %d\n", field_count);
        printf("  First 10 fields:\n");
        for (int i = 0; i < field_count && i < 10; i++) {
            printf("    %d. %s\n", i + 1, field_names[i]);
        }
        ms2_free_fields(field_names, field_count);
    } else {
        fprintf(stderr, "ERROR: Failed to list fields\n");
    }

    /* Disconnect */
    printf("\nDisconnecting...\n");
    ms2_disconnect(client);
    printf("Done.\n");

    return 0;
}

File: work-ms2d-daemon/ms2d/clients/ms2_client_example.js
#!/usr/bin/env node

/**
 * MS2D Client Example
 * 
 * Demonstrates usage of the MS2Client module to connect to ms2d daemon
 * and retrieve sensor values.
 * 
 * Usage: node ms2_client_example.js [socket_path]
 * Example: node ms2_client_example.js /tmp/ms2d.sock
 */

const { MS2Client } = require('./ms2_client');

async function main() {
  // Get socket path from command line or use default
  const socketPath = process.argv[2] || '/tmp/ms2d.sock';

  console.log(`Connecting to ms2d daemon at: ${socketPath}`);

  const client = new MS2Client(socketPath);

  try {
    // Connect to daemon
    await client.connect();
    console.log('✓ Connected to daemon\n');

    // Get daemon status
    console.log('--- Daemon Status ---');
    const status = await client.getStatus();
    console.log(JSON.stringify(status, null, 2));
    console.log();

    // Get list of available fields
    console.log('--- Available Fields ---');
    const fields = await client.listFields();
    console.log(`Total fields: ${fields.length}`);
    console.log(`First 10 fields: ${fields.slice(0, 10).join(', ')}`);
    console.log();

    // Get a single field value (RPM)
    console.log('--- Single Field Value (RPM) ---');
    const rpm = await client.getValue('rpm');
    console.log(JSON.stringify(rpm, null, 2));
    console.log();

    // Get multiple field values
    console.log('--- Multiple Field Values ---');
    const values = await client.getValues(['rpm', 'batteryVoltage', 'tps']);
    console.log(JSON.stringify(values, null, 2));
    console.log();

    // Get all field values
    console.log('--- All Field Values (first 5) ---');
    const allValues = await client.getAll();
    console.log(`Total values: ${allValues.values.length}`);
    console.log('First 5 values:');
    console.log(JSON.stringify(allValues.values.slice(0, 5), null, 2));
    console.log();

    // Disconnect
    await client.disconnect();
    console.log('✓ Disconnected from daemon');

  } catch (err) {
    console.error('✗ Error:', err.message);
    process.exit(1);
  }
}

main();
